src/manifest/owned/mod.rs

```diff
@@ -0,0 +1,412 @@
+
+    /// Returns the element type of a specific element.
+    fn element_type(&self) -> Self::ElementType;
+
+    /// Attempts to construct an `Element` of the given type from encoded
+    /// `bytes`.
+    fn from_bytes(ty: Self::ElementType, bytes: &[u8]) -> Result<Self, Error>;
+
+    /// Attempts to encode this `Element` into bytes, using the given
+    /// padding byte as "filler".
+    fn to_bytes(&self, padding_byte: u8) -> Result<Vec<u8>, Error>;
+}
+
+/// A heap-allocated PFM.
+///
+/// See [`manicore::manifest::pfm`] for lazy parsing out of flash.
+///
+/// [`manticore::manifest::pfm`]: ../pfm/index.html
+pub type Pfm = Container<self::pfm::Element>;
+
+/// A heap-allocated Cerberus manifest, represented as a tree structure.
+///
+/// Prefer to access this type through one of the provided type aliases,
+/// instead:
+/// - Platform Firmware Manifest: [`Pfm`](type.Pfm.html)
+/// - Component Firmware Manifest: NYI
+/// - Platform Configuration Descriptor: NTI
+#[derive(Clone, PartialEq, Eq, Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct Container<E> {
+    /// The metadata for this manifest.
+    #[cfg_attr(feature = "serde", serde(flatten))]
+    pub metadata: Metadata,
+
+    /// The root elements of this manifest.
+    pub elements: Vec<Node<E>>,
+}
+
+/// An element node within a manifest [`Container`].
+///
+/// [`Container`] struct.Container.html
+#[derive(Clone, PartialEq, Eq, Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct Node<E> {
+    /// Whether this node should be hashed.
+    #[cfg_attr(
+        feature = "serde",
+        serde(
+            default = "crate::serde::default_to_true",
+            skip_serializing_if = "crate::serde::skip_if_true"
+        )
+    )]
+    pub hashed: bool,
+
+    /// The element that this node contains.
+    #[cfg_attr(feature = "serde", serde(flatten))]
+    pub element: E,
+
+    /// This element's children, if any are present.
+    #[cfg_attr(
+        feature = "serde",
+        serde(default = "Vec::new", skip_serializing_if = "Vec::is_empty")
+    )]
+    pub children: Vec<Node<E>>,
+}
+
+/// A parsing result.
+///
+/// See [`Container::parse()`].
+///
+/// [`Container::parse()`]: struct.Container.html#method.parse
+pub struct Parse<E> {
+    /// The container resulting from the parse.
+    pub container: Container<E>,
+
+    /// Whether the container had a bad signature.
+    pub bad_signature: bool,
+
+    /// Whether the TOC hash was bad.
+    pub bad_toc_hash: bool,
+
+    /// A list of indices of TOC entries whose elements had bad hashes.
+    pub bad_hashes: Vec<usize>,
+}
+
+impl<E: Element> Container<E> {
+    /// Parses a `Container` out of `bytes`.
+    ///
+    /// Note that if a cryptographic operation fails, `Ok` will still be returned;
+    /// the resulting value will indicate which operations failed in addition to
+    /// containing the parsed container.
+    pub fn parse(
```

Comment:
How come you can't reuse the parsing code? Having N>1 parsers can lead to a lot of issues, many of which have security implications.

Reply:
This is currently a stopgap, due to the way the commits are layered. The idea is actually to use the "unowned" Container/PFM parser to walk an encoded manifest and dump it into the heap, but that requires re-doing the API for Container/PFM that would have made this PR even bigger than it has to be[1]. Look out for a PR that will eliminate this parser.

Reply:
Got it, thanks!